src/ray/gcs/gcs_server/gcs_server.cc

```diff
@@ -599,5 +604,33 @@ void GcsServer::PrintAsioStats() {
   }
 }
 
+void GcsServer::DetectHang() {
```

Comment:
let's create a helper class (such as AsioHangDetector) so this logic could be reused by any asio::service.